contracts/root/StakeManager.sol

```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+  uint256 public totalStake = 0;
+  uint256 public currentEpoch = 1;
+  uint256 public totalEpoch;
 
-  // The randomness seed of the epoch.
-  // This is used to determine the proposer and the validator pool
-  bytes32 public epochSeed = keccak256(abi.encodePacked(block.difficulty + block.number + now));
+  //Todo: dynamically update
+  uint256 public minStakeAmount = 1;  // ETH
+  uint256 public minLockInPeriod = 100; //(unit epochs)
+  uint256 public stakingIdCount = 0;  // just a counter/index to map it with PQ w/address
 
-  //
-  // Events
-  //
+  enum ValidatorStatus { WAITING, VALIDATOR, UNSTAKING } // need update 
 
-  event ThresholdChange(uint256 newThreshold, uint256 oldThreshold);
+  struct Staker {
+    uint256 epoch;  // init 0 
+    uint256 amount;
+    bytes data;
+    ValidatorStatus status;
+    uint256 stakingId; 
+  }
 
-  // ERC900
-  event Staked(address indexed user, uint256 amount, uint256 total, bytes data);
-  event Unstaked(address indexed user, uint256 amount, uint256 total, bytes data);
+  address[] exiterList;
+  address[] stakersList;
 
-  //
-  // Constructor
-  //
+  mapping (address => Staker) stakers; 
+  mapping (uint256 => address) stakingIdToAddress;
 
-  constructor (address _token) public {
+  constructor(address _token) public {
     require(_token != 0x0);
     tokenObj = ERC20(_token);
+    stakeQueue = new PriorityQueue();
   }
 
-  //
-  // Modifiers
-  //
-
   // only staker
   modifier onlyStaker() {
     require(totalStakedFor(msg.sender) > 0);
     _;
   }
 
-  //
-  // ERC900 implementation
-  //
+  
+  function _priority(address user, uint256 amount, bytes data) private view returns (uint256) {
+    // return amount.mul(10).add(currentEpoch.mul(1000).add(amount.mul(100).div(user.balance)));
+    return amount;
+  }
 
   function stake(uint256 amount, bytes data) public {
+    // no second time staking 
+    // maybe restrict entry after n staker
+    require(stakers[msg.sender].epoch == 0);
     stakeFor(msg.sender, amount, data);
   }
 
   function stakeFor(address user, uint256 amount, bytes data) public onlyWhenUnlocked {
-    require(amount > 0);
-
-    // transfer tokens to stake manager
-    require(tokenObj.transferFrom(msg.sender, address(this), amount));
-
-    // check staker is present or not
-    if (stakers[user] == 0) {
-      // actual staker cannot be on index 0
-      if (stakeList.length == 0) {
-        stakeList.push(Stake({
-          amount: 0,
-          staker: address(0x0)
-        }));
-      }
-
-      // add new stake
-      stakeList.push(Stake({
-        amount: 0,
-        staker: user
-      }));
-      stakers[user] = stakeList.length - 1;
+    require(amount >= minStakeAmount); 
+
+    // actual staker cannot be on index 0
+    if (stakersList.length == 0) {
+      stakersList.push(address(0x0));
+      stakers[address(0x0)] = Staker(0, 0, new bytes(0), ValidatorStatus.WAITING, stakingIdCount);
+      stakingIdToAddress[stakingIdCount] = address(0x0);
```

Comment:
Do we need stakingIdToAddress?

Reply:
yes it's for PQ's uint to address mapping, maybe will modify PQ to support address or use Balanced BST to improve on some of the things.